crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2660,6 +2660,55 @@ impl<'db> TypeInferenceBuilder<'db> {
                 self.infer_binary_type_comparison(left, op, KnownClass::Str.to_instance(self.db))
             }
 
+            (Type::Tuple(lhs), Type::Tuple(rhs)) => {
+                // Note: This only works on heterogeneous tuple types.
+                let lhs_elements = lhs.elements(self.db).as_ref();
+                let rhs_elements = rhs.elements(self.db).as_ref();
+
+                match op {
+                    ast::CmpOp::Eq
+                    | ast::CmpOp::NotEq
+                    | ast::CmpOp::Lt
+                    | ast::CmpOp::LtE
+                    | ast::CmpOp::Gt
+                    | ast::CmpOp::GtE => {
+                        self.infer_lexicographic_type_comparison(lhs_elements, op, rhs_elements)
+                    }
+                    ast::CmpOp::In | ast::CmpOp::NotIn => {
+                        // - `[ast::CmpOp::In]`: returns `true` if the element is definitely in the tuple, otherwise `bool`
+                        // - `[ast::CmpOp::NotIn]`: returns `false` if the element is definitely in the tuple, otherwise `bool`
+                        if rhs_elements.iter().any(|ty| {
+                            let eq_result_ty = self.infer_binary_type_comparison(
+                                Type::Tuple(lhs),
+                                ast::CmpOp::Eq,
+                                *ty,
+                            );
+
+                            matches!(eq_result_ty, Some(Type::BooleanLiteral(true)))
+                        }) {
+                            Some(Type::BooleanLiteral(op.is_in()))
+                        } else {
+                            Some(KnownClass::Bool.to_instance(self.db))
+                        }
```

Comment:
I think in this case there is an implicit invariant that `eq_result_ty` should never be `None`, because equality comparison is supported between all types (it never raises `TypeError` at runtime). We could assert that, to be sure. Because  if somehow a `None` return were possible, we would want to propagate it so we get an unsupported operation diagnostic message.